### ISSUE TITLE
bug in config subcategory usage

### ISSUE THREAD

**Issue by sueloverso** [wiredtiger/wiredtiger#1879]
I came across a problem with the way we generate structures for subcategories the other day.  I encountered it with our new `encryption` category.  For our example, we have an `encryption` subcategory available on both `wiredtiger_open` (i.e. defined in the `connection_runtime_config` area of `api_data.py`) and `WT_SESSION::create` (defined in `file_config`).

In our example, the `WT_SESSION::create` definition had 2 items, name and keyid.  The definition for `wiredtiger_open` had 3 items, name, keyid and passwd.  (In this case the two categories overlap, but one could easily imagine a generically named category used multiple times that may not overlap.)

When running the scripts, the definition for `WT_SESSION::create` is the one that "wins" and we end up with a structure that looks like:
```
static const WT_CONFIG_CHECK confchk_encryption_subconfigs[] = {
        { "keyid", "string", NULL, NULL, NULL, 0 },
        { "name", "string", NULL, NULL, NULL, 0 },
        { NULL, NULL, NULL, NULL, NULL, 0 }
};
```
Then when `passwd` is used in `wiredtiger_open` we segfault.

I'm able to make progress by simply changing the category name in `WT_SESSION::create`.

**Comment by keithbostic** [comment]
This is non-trivial to change, there's no "name" here to use as the unique identifier.

@michaelcahill, you know this code a lot better than I do, do you have any ideas?

If we leave this unchanged, we should at least modify `api_config.py` to detect the problem.

**Comment by keithbostic** [comment]
On second thought, we could probably pull out the subconfigs into their own class and give them a name, and then reference those names from inside a Config?

**Comment by sueloverso** [comment]
Here are a few random thoughts:
1.  I agree detecting the problem would be a minimum to avoid a segfault.  However, it isn't ideal because in this case these two subcategories are related and naming them differently is detrimental and confusing for the user.  Right now I have `encryption=(...)` in `wiredtiger_open` and `encrypt=(...)` for `WT_SESSION::create`.
2.  How difficult or gross would it be to create a superset of all uses?  In my case it would end up with what I need for `wiredtiger_open`.  This is only palatable for related ones but someone should think hard about a category named the same used for unrelated things.

**Comment by ddanderson** [comment]
I think this can be fixed by naming the subconfigs to include the name of the referencing config: `confchk_session_create_encryption_subconfig`, rather than just `confchk_encryption_subconfig`.  These names are just used internally to `config_def.c`.  I don't think any changes beyond the script. I'll give a shot at this.

**Comment by michaelcahill** [comment]
Closed by the merge of #1882.
